### PR TITLE
Upgrade to tendermint v0.35.4 and cosmos-sdk v0.46.0

### LINKED ISSUE #376: Upgrade to celestia-core `v1.2.0-tm-v0.35.4`
We want to upgrade to the latest version of celestia-core `v1.2.0-tm-v0.35.4` https://github.com/celestiaorg/celestia-core/issues/567, which means that we will also have to upgrade to a branch of our fork of the cosmos-sdk that also uses that version of celestia-core https://github.com/celestiaorg/cosmos-sdk/issues/130 .

### PR DESCRIPTION
**Pull request by evan-forbes**
## Description

This PR updates the application to tendermint v0.35.4 and the cosmos-sdk to v0.46.0. While there are a wide variety of changes that included in this PR, each is needed to accommodate the latest version of the cosmos-sdk

blocked by https://github.com/celestiaorg/celestia-core/pull/747
closes #376

related to https://github.com/celestiaorg/celestia-app/issues/291

### CODE REVIEW

File: app/encoding/encoding.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// This is provided for compatibility between protobuf and amino implementations.
+type EncodingConfig struct {
+	InterfaceRegistry codectypes.InterfaceRegistry
+	Codec             codec.Codec
+	TxConfig          client.TxConfig
+	Amino             *codec.LegacyAmino
+}
+
+// MakeEncodingConfig creates an encoding config for the app.
+func MakeEncodingConfig(regs ...InterfaceRegister) EncodingConfig {
```
Comment (evan-forbes) [inline app/encoding/encoding.go:23]: Here we are refactoring the encoding config into its own package, and instead of relying on a the [`ModuleBasics`](https://github.com/celestiaorg/celestia-app/blob/3c8376459a9c59a3a96490fc990bbafa9346cd5f/app/app.go#L125)  defined in the app, we pass a function that registers interfaces. By passing that same [method](https://github.com/celestiaorg/celestia-app/blob/3c8376459a9c59a3a96490fc990bbafa9346cd5f/x/payment/types/codec.go#L21-L41) in a module, we can use a single `MakeEncodingConfig` function everywhere in this repo

this also allows us to get rid of all the `cosmoscmd` starport code

File: cmd/celestia-appd/main.go [hunk 2]
```diff
@@ -1,58 +1,15 @@
...
-		app.AccountAddressPrefix,
-		app.DefaultNodeHome,
-		app.Name,
-		app.ModuleBasics,
-		appBuilder,
-		// this line is used by starport scaffolding # root/arguments
-	)
-	if err := svrcmd.Execute(rootCmd, app.DefaultNodeHome); err != nil {
+	rootCmd := NewRootCmd()
+	if err := svrcmd.Execute(rootCmd, "CELESTIA", app.DefaultNodeHome); err != nil {
```
Comment (evan-forbes) [inline cmd/celestia-appd/main.go:12]: We replace more starport code with our own infrastructure.

File: cmd/celestia-appd/root.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+	tmcmds "github.com/tendermint/tendermint/cmd/tendermint/commands"
+	tmcfg "github.com/tendermint/tendermint/config"
+	tmcli "github.com/tendermint/tendermint/libs/cli"
+	"github.com/tendermint/tendermint/libs/log"
+	dbm "github.com/tendermint/tm-db"
+)
+
+// NewRootCmd creates a new root command for celestia-appd. It is called once in the
+// main function.
+func NewRootCmd() *cobra.Command {
```
Comment (evan-forbes) [inline cmd/celestia-appd/root.go:41]: This command can probably be cleaned up and more organized. I elected to do this in a different PR for time purposes, but if we want I'm happy to refactor to something cleaner and more organized

File: testutil/test_app.go [hunk 4]
```diff
@@ -163,10 +197,103 @@ func generateKeyring(t *testing.T, accts ...string) keyring.Keyring {
 	return kb
 }
 
+// SetupWithGenesisValSet initializes GenesisState with a single validator and genesis accounts
+// that also act as delegators.
+func GenesisStateWithSingleValidator(t *testing.T, testApp *app.App) (app.GenesisState, *tmtypes.ValidatorSet) {
```
Comment (evan-forbes) [inline testutil/test_app.go:202]: This might fix the issue @renaynay was running into with using a test application in celestia-node, as the network could not be setup due to improper initialization of the staking state/validator set

File: go.mod [hunk 5]
```diff
@@ -4,128 +4,158 @@ go 1.17
...
 
 replace (
 	github.com/99designs/keyring => github.com/cosmos/keyring v1.1.7-0.20210622111912-ef00f8ac3d76
-	github.com/cosmos/cosmos-sdk => github.com/celestiaorg/cosmos-sdk v1.0.1-sdk-v0.44.2
+	github.com/cosmos/cosmos-sdk => github.com/celestiaorg/cosmos-sdk v1.1.0-sdk-v0.46.0.0.20220510181438-f94c672be3d9
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
-	github.com/tendermint/tendermint => github.com/celestiaorg/celestia-core v1.0.1-tm-v0.34.16
-	google.golang.org/grpc => google.golang.org/grpc v1.33.2
+	github.com/tendermint/tendermint v0.35.4 => github.com/celestiaorg/celestia-core v1.2.0-tm-v0.35.4.0.20220512001247-7760c1030e01
 )
```
Comment (evan-forbes) [inline go.mod:161]: we still need to switch to a proper release after [#747](https://github.com/celestiaorg/celestia-core/pull/747)

also, we can finally use an updated version of grpc!!! v1.33 -> v1.46 :tada:

**Comment by adlerjohn** [comment]
Converting to draft until unblocked

File: app/app.go [hunk 6]
```diff
@@ -120,15 +130,15 @@ var (
...
 		mintModule{},
 		distr.AppModuleBasic{},
-		gov.NewAppModuleBasic(getGovProposalHandlers()...),
+		gov.NewAppModuleBasic(getGovProposalHandlers()),
 		params.AppModuleBasic{},
 		crisisModule{},
 		slashing.AppModuleBasic{},
 		feegrantmodule.AppModuleBasic{},
-		ibc.AppModuleBasic{},
+		// ibc.AppModuleBasic{},
```
Comment (liamsi) [inline app/app.go:138]: Can we add a comment why this is commented out? (I assume because IBC wasn't fully updated yet?)

File: app/app.go [hunk 7]
```diff
@@ -120,15 +130,15 @@ var (
...
 		upgrade.AppModuleBasic{},
 		evidence.AppModuleBasic{},
-		transfer.AppModuleBasic{},
+		// transfer.AppModuleBasic{},
```
Comment (liamsi) [inline app/app.go:141]: Same as above.

File: app/app.go [hunk 8]
```diff
@@ -143,13 +153,12 @@ var (
 		minttypes.ModuleName:           {authtypes.Minter},
 		stakingtypes.BondedPoolName:    {authtypes.Burner, authtypes.Staking},
 		stakingtypes.NotBondedPoolName: {authtypes.Burner, authtypes.Staking},
-		ibctransfertypes.ModuleName:    {authtypes.Minter, authtypes.Burner},
+		// ibctransfertypes.ModuleName:    {authtypes.Minter, authtypes.Burner},
 		// this line is used by starport scaffolding # stargate/app/maccPerms
```
Comment (liamsi) [inline app/app.go:157]: Can delete this comment?

File: app/app.go [hunk 9]
```diff
@@ -143,13 +153,12 @@ var (
 		minttypes.ModuleName:           {authtypes.Minter},
 		stakingtypes.BondedPoolName:    {authtypes.Burner, authtypes.Staking},
 		stakingtypes.NotBondedPoolName: {authtypes.Burner, authtypes.Staking},
-		ibctransfertypes.ModuleName:    {authtypes.Minter, authtypes.Burner},
+		// ibctransfertypes.ModuleName:    {authtypes.Minter, authtypes.Burner},
```
Comment (liamsi) [inline app/app.go:156]: same as above?

File: app/app.go (hunk 7 above)
Comment (liamsi) [reply app/app.go:141]: Actually maybe we should just create an issue and somehow make sure we won't miss those places.

File: app/export.go [hunk 10]
```diff
@@ -110,14 +113,24 @@ func (app *App) prepForZeroHeightGenesis(ctx sdk.Context, jailAllowedAddrs []str
 		feePool.CommunityPool = feePool.CommunityPool.Add(scraps...)
 		app.DistrKeeper.SetFeePool(ctx, feePool)
 
-		app.DistrKeeper.Hooks().AfterValidatorCreated(ctx, val.GetOperator())
+		err := app.DistrKeeper.Hooks().AfterValidatorCreated(ctx, val.GetOperator())
+		if err != nil {
+			panic(err)
```
Comment (liamsi) [inline app/export.go:118]: What is the rationale for panicing here? Can this be documented too?

File: cmd/celestia-appd/main.go (hunk 2 above)
Comment (liamsi) [inline cmd/celestia-appd/main.go:12]: Should the  "CELESTIA" string be in a const? (e.g. `envPrefix`)

File: cmd/celestia-appd/root.go [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+		serverconfig.Config
+	}
+
+	// Optionally allow the chain developer to overwrite the SDK's default
+	// server config.
+	srvCfg := serverconfig.DefaultConfig()
+	srvCfg.API.Enable = true
+	srvCfg.StateSync.SnapshotInterval = 1500
+	srvCfg.StateSync.SnapshotKeepRecent = 2
+	srvCfg.MinGasPrices = fmt.Sprintf("0%s", app.BondDenom)
```
Comment (liamsi) [inline cmd/celestia-appd/root.go:104]: Can you document these?

File: contrib/single-node.sh [hunk 12]
```diff
@@ -18,6 +18,7 @@ sed -i 's#"tcp://127.0.0.1:26657"#"tcp://0.0.0.0:26657"#g' ~/.celestia-app/confi
 sed -i 's/timeout_commit = "5s"/timeout_commit = "1s"/g' ~/.celestia-app/config/config.toml
 sed -i 's/timeout_propose = "3s"/timeout_propose = "1s"/g' ~/.celestia-app/config/config.toml
```
Comment (liamsi) [inline contrib/single-node.sh:19]: Are we intentionally overwriting this for single node networks?

File: go.mod [hunk 13]
```diff
@@ -4,128 +4,159 @@ go 1.17
 
 require (
 	github.com/celestiaorg/nmt v0.8.0
-	github.com/cosmos/cosmos-sdk v0.44.3
-	github.com/cosmos/ibc-go v1.2.2
+	github.com/cosmos/cosmos-sdk v0.46.0-beta2.0.20220418184507-c53157dd63f6
```
Comment (liamsi) [inline go.mod:7]: There is no tagged release here yet?

File: testutil/test_app.go [hunk 14]
```diff
@@ -1,90 +1,124 @@
...
 }
 
-// SetupTestApp initializes a celestia-app application with a funded account
-func SetupTestApp(t *testing.T, addr sdk.AccAddress) *app.App {
+// DefaultConsensusParams defines the default Tendermint consensus params used in
+// SimApp testing.
+var DefaultConsensusParams = &tmproto.ConsensusParams{
+	Block: &tmproto.BlockParams{
+		MaxBytes: 200000,
+		MaxGas:   2000000,
```
Comment (liamsi) [inline testutil/test_app.go:47]: Not sure why we would keep the blocks so small in tests 🤔

**Review by liamsi — COMMENTED** [review]
Left a bunch of comments. Looks really good besides a few nits and questions. I did not test this manually.

File: app/export.go (hunk 10 above)
Comment (evan-forbes) [reply app/export.go:118]: mainly because the function that the `IterateValidators` method relies on as input doesn't handle errors. Also, errors here should only be because of invalid validator data was committed to, which we don't know how to handle anyway.

File: cmd/celestia-appd/root.go (hunk 11 above)
Comment (evan-forbes) [reply cmd/celestia-appd/root.go:109]: sure, but I'm unsure what document with as I didn't actually have any very good reason for changing these other than the default for snapshot interval is to not take any snapshots, which would reduce ssd space used by the node, but would very bad for state syncing as most nodes would not store any backups.

the default min gas price is actually insecure, because it allows for ddos of p2p network with free txs. The only reason it was changed, because the default uses `stake` instead of `celes`.

created an issue for further discussion https://github.com/celestiaorg/celestia-app/issues/413

I'm also happy to change these values back to the default too, left a comment for snapshot intervals
https://github.com/celestiaorg/celestia-app/pull/399/commits/b3eb11219eb23f6203675b7e7fefeaf80895ee95

File: contrib/single-node.sh (hunk 12 above)
Comment (evan-forbes) [reply contrib/single-node.sh:19]: yeah, tbh, I don't remember why or when we added this

File: cmd/celestia-appd/main.go (hunk 2 above)
Comment (evan-forbes) [reply cmd/celestia-appd/main.go]: good idea, https://github.com/celestiaorg/celestia-app/pull/399/commits/740ad0045ce96817e5353b972ffed6f088aadc1f

File: app/app.go (hunk 8 above)
Comment (evan-forbes) [reply app/app.go]: yes! removed all of them

https://github.com/celestiaorg/celestia-app/pull/399/commits/f201a83a9ee61c97d5216b4132ada95113813265

File: app/app.go (hunk 7 above)
Comment (evan-forbes) [reply app/app.go:142]: :+1: good idea

https://github.com/celestiaorg/celestia-app/issues/412

https://github.com/celestiaorg/celestia-app/pull/399/commits/3a58c64fdc45e87b419df7dff13b32c5cf991fcf

**Comment by liamsi** [comment]
Surprised to see merge conflicts now. I think we did not change anything on main?

**Comment by evan-forbes** [comment]
> Surprised to see merge conflicts now. I think we did not change anything on main?

we merged #224, so this makes sense. I tried to quickly merge last night, but ran into an issue and haven't time to finish debugging.

File: go.mod (hunk 13 above)
Comment (adlerjohn) [reply go.mod:7]: Still waiting on this one

Comment (evan-forbes) [reply go.mod:7]: whoops, thanks for the reminder!! https://github.com/celestiaorg/celestia-app/pull/399/commits/17379f74ca7178e5bc0dfe0d284ad1903fc25c57

File: testutil/test_app.go (hunk 14 above)
Comment (evan-forbes) [reply testutil/test_app.go:47]: so far, this test is only used for a few txs, as its only for unit tests. We might need to change this in the near future, depending on when and how we add more tests with large messages and full mempools

**Review by adlerjohn — APPROVED** [review]
utACK

**Review by liamsi — APPROVED** [review]
